aip/0002.md

```diff
@@ -11,45 +11,54 @@ redirect_from:
   - /0002
 ---
 
-# AIP Numbering
+# AIP Identifiers
 
 The AIP system provides a mechanism to index and have a single source of truth
 for API Improvement Proposals, as well as iterate on them collaboratively and
 transparently.
 
-This document describes the AIP numbering system.
+Multiple companies and teams have adopted the AIP system, with the mutual goal
+that the identifiers for common concepts should be shared between each
+company's AIP program to provide comparability.
 
 ## Assigning AIP Numbers
 
-The AIP editors (see [AIP-1](./0001.md)) are responsible for assigning a number
-to each AIP when it is accepted as a draft for review. Importantly, _all_ AIPs
-have numbers, not just approved ones. The [AIP Index](/) clearly delineates
-which AIPs are approved and binding and which are under discussion.
+The AIP editors (see [AIP A-1](./0001.md)) are responsible for assigning a
+unique identifier to each AIP when it is accepted as a draft for review. For
+AIPs expected to be common between companies, a common prefix (either one or
+two letters) and one number are assigned, and shared across the system. AIPs
+specific to a single company are given a three-letter prefix instead.
+Three-letter prefixes are only guaranteed to be unique within a given company.
 
-The editors may decide to reserve a specific block of numbers for groups of
-AIPs that are related in some way (for example, that are only scoped to a
-specific subset of APIs).
-
-Beyond this, AIP numbers are assigned arbitrarily. In general, the editors will
-take the next AIP number off of the stack to assign to a draft AIP, but
-occasionally may use a special/joke number if useful for mnemonic or other
-reasons.
+Importantly, _all_ AIPs have identifiers, not just approved ones. The
+[AIP Index](/) clearly delineates which AIPs are approved and binding and which
+are under discussion.
 
 ## AIP Blocks
 
-Currently recognized blocks of AIP numbers are:
-
-### Generally Applicable
+Common AIPs are organized into the following over-arching categories:
 
-- **1-99:** Reserved for meta-AIPs (generally process-related).
-- **100-999:** General API design guidance
+- **M:** Meta-AIPs (AIPs about AIPs)
+- **RD:** Resource design
+- **SM:** Standard methods
+- **CM:** Custom methods
+- **F:** Fields
+- **DP:** Design Patterns
+- **C:** Compatibility
+- **S:** File Structure
+- **A:** Annotations
+- **P:** Process
 
```

Comment:
```suggestion
### Sub-AIPs
There may be times when an AIP adopter will want to adopt an AIP and want to provide additional requirements or clarifications. Rather than turning an AIP into a custom AIP by adding these additional standards, or to prevent the AIP from becoming bloated, AIP adopters can add sub AIPs using a period (`.`) after the main AIP identifier.

Some sub AIPs are standard as well. For example, Field Types are defined as sub-AIPs of Field Types to ensure that as new Field Type standards are added to the AIP standard that they will always be grouped under the same AIP, while still being individually extensible and can be overridden.

Sub-AIPs can also have their own Sub-AIPs, defined the same way.

### Example (SF):
The following is an example of the above AIP rules applied to the standard AIPs definition proposed in [#402](https://github.com/googleapis/aip/issues/98#issuecomment-565649006). It is presented here to allow visualization of how an AIP collection might look:

G-100: Design Philosophy/Mission Statement/Values/etc.
G-200: Versioning
G-300: Documentation
G-400: Bad API Precedent

P-100: API Lifecycle
P-200: Change Management

T-100: General
T-200: Metadata (Headers)
T-300: Requests
T-400: Responses
T-400.10: Errors
T-400.20: Pagination
T-400.30: Partial

RD-100: Resource Design - General
RD-200: Names
RD-300: Types
RD-400: Caching
RD-500: Singleton
RD-600: Unreachable

F-100: Fields -  General
F-200: Names
F-300: Types
F-400.10: Types - Quantities
F-500.20: Types - Enumerations
F-600.30: Types - Time and Duration
F-700.40: Types - Standardized Codes

SM-100: Standard Methods - General
SM-200: Retrieval - General
SM-200.10: Retrieval - Single Resource
SM-200.20: Retrieval - Resource Collection
SM-300: Update - General
SM-300.10: Update - Single Resource
SM-300.20: Update - Resource Collection
SM-400: Delete - General
SM-400.10: Delete - Single Resource
SM-400.20: Delete - Resource Collection
SM-500: Create - General
SM-500.10: Create - Single Resource
SM-500.20: Create - Resource Collection

CM-100: Custom Methods - General
```

Reply:
I'm adding this as a suggestion recognizing that it is rather large and might be easier to add to the PR and allow folks to comment directly on the code (if that works, I've not tried suggestions before).

This is somewhat based on my experience migrating our Standards doc to AIPs. I organized my AIPs based on my proposal in #402. What I found was that I liked the idea of introducing a category, addressing cross cutting concerns, and then diving deeper into examples.

For example, my implementation of Fields looks like this:
```
3000 - Fields
3100 - Field Names
3200 - Field Types
3210 - Field Types: Quantities
3220 - Field Types: Enumerations
3230 - Field Types: Time and Duration
3240 - Field Types: Standardized Codes 
```
You can see how even using four digits, I am already limiting the extensibility of my AIP definition.

Reply:
There is some precedent for this idea (law works this way, for example). I am...hesitant though. I am not sure where the sweet spot is between too much structure and not enough.

Reply:
Ha, I read this and thought "Oh, he's really going to hate the '. extension' idea in the next section then" :D 

Personally I worry about the '. extension' approach being too much structure, but that's not until the next section.

Just to make sure we are on the same page though, what part do you think is too much structure in the example above?

Reply:
Right now, it is useful that we can _say_, out loud, an AIP citation and everyone knows exactly what to do. I have a lot of conversations in videoconference that go something like:

> "Hey Luke, we wanted to know what to do if we have this one field, and it is kind of expensive, and it would be really nice if..."
> "AIP-157."
> "Great, thanks!"

In other words, while it is definitely true that we link these all the time, it is also true that the current structure allows for every simple and readily understandable citation. Heck, a lot of times, the person does not need to even write it down.

By that logic, I think there is some loss if we start adding multiple forms of punctuation and such that people have to think about. This aspect will have gotten a bit worse if I start saying, "AIP SM dash 400 dot 20".

Now, to be clear, maybe we gain more than we lose. But we definitely do lose _something_.

At one level, I think that loss is much less. Heck, let's say we styled it something like AIP RD-4 -- I could even set up redirects so that you can get the punctuation wrong and it just does not matter -- `/rd-4` and `/rd4` and `/rd_4` and `/rd.4` can all just work if we want them to. I think that means I can still just tell people "AIP RD 4" and it will likely mostly work (with caveats for having to say "Romeo Delta" sometimes because B and D sound alike, but meh).

Like I said, definitely open to more structure, but I want to be careful not to introduce _too_ much. I am not sure if what you are proposing is too much or not, but I wanted to explain what I view the tradeoff as being.

Reply:
> Now, to be clear, maybe we gain more than we lose. But we definitely do lose something.

I understand, and I don't think you're wrong.

I mostly wanted to clarify that it was the punctuation you were referring to when you meant structure.

I don't think you are necessarily against having AIPs (regardless of punctuation) that are some form of "sub" AIP of another one (please confirm). In my example above I use 3000 for fields, and 3100 for field names. 3100 could be thought of as a type of sub-aip of 3000.

And I think I will leave my comment on this one there for now. I think that is the question we probably need to consider before we can discuss how that should be represented. What do you think?

Reply:
> I don't think you are necessarily against having AIPs (regardless of punctuation) that are some form of "sub" AIP of another one (please confirm). 

Confirmed, if we can find a mechanism that scales well and is sufficiently easy to communicate. (I leave "sufficiently easy to communicate" intentionally vague.)

> I think that is the question we probably need to consider before we can discuss how that should be represented. What do you think?

Agree wholeheartedly.

Reply:
I think we agree we will have sub-aips, but I think that's a larger conversation. I created Issue #452 as a placeholder for this discussion.

Now that I wrote it and am looking at what I wrote, perhaps we actually need to better define what a sub-aip is. My examples have me questioning what they should be.